README.md
# Octopus Deploy Integration Pack

This integration pack allows you to integrate with
[Octopus Deploy](http://octopusdeploy.com/),
deployment automation system for .NET applications.

## Actions

The following actions are supported:

### Projects

* Get Projects - `list_projects`
* Get Deployments - `list_deployments`

### Releases

* Create a new release - `create_release`
* Deploy a release to an environment - `deploy_release`
* Get a list of releases for a project - `list_releases`

### Machines (tentacles)

* Add a new machine to an environment(s) - `add_machine`

## Sensors

* Detect a new release being created - `new_release_sensor`
* Detect a new deployment being created - `new_deployment_sensor`

## Configuration

Copy the example configuration in [octopusdeploy.yaml.example](./octopusdeploy.yaml.example)
to `/opt/stackstorm/configs/octopusdeploy.yaml` and edit as required. It must contain:

* `api_key` - an API key generated in Octopus for your user http://docs.octopusdeploy.com/display/OD/How+to+create+an+API+key 
* `host` - the host name of your Octopus server e.g. octopus.mydomain.com
* `port` - the port your API service is running on, 443 by default

Read this pack tutorial for more information: https://stackstorm.com/2015/10/01/octopusdeploy-integration-with-stackstorm/


CHANGES.md
# Change Log

## 0.1.4

- Python 3 fixups
- Add explicit support for Python 2 and 3

## 0.1.3

- Minor linting fix


actions/lib/actions.py
try:
    import requests
except ImportError:
    message = ('Missing "requests", please install it using pip:\n'
               'pip install requests')
    raise ImportError(message)

try:
    import json
except ImportError:
    message = ('Missing "json", please install it using pip:\n'
               'pip install json')
    raise ImportError(message)

from octopus_error import OctopusError
from st2common.runners.base_action import Action

__all__ = [
    'OctopusDeployAction',
]


class OctopusDeployAction(Action):
    def __init__(self, config):
        super(OctopusDeployAction, self).__init__(config)
        self.client = self._init_client()

    def _init_client(self):
        api_key = self.config['api_key']
        host = self.config['host']
        port = self.config['port']
        return OctopusDeployClient(api_key=api_key, host=host, port=port)

    def _build_uri(self):
        # big assumption but it'll cover 99% case,
        # as octopus runs https by default
        start = "http://" if self.client.port == 80 else "https://"
        return start + self.client.host + ":" + str(self.client.port) + "/api/"

    def make_post_request(self, action, payload):
        response = requests.post(self._build_uri() + action,
                                 data=json.dumps(payload), verify=False,
                                 headers=self.client.headers)
        if response.status_code != 200:
            raise OctopusError(response.text, response.status_code)
        return response.json()

    def make_get_request(self, action, params=None):
        response = requests.get(self._build_uri() + action,
                                verify=False,
                                params=params,
                                headers=self.client.headers)
        if response.status_code != 200:
            raise OctopusError(response.text, response.status_code)
        return response.json()


class OctopusDeployClient(object):
    def __init__(self, api_key, host, port):
        self.api_key = api_key
        self.host = host
        self.port = port
        self.headers = {'X-Octopus-ApiKey': self.api_key,
                        'Content-type': 'application/json',
                        'Accept': 'text/plain'}


sensors/new_release_sensor.py
import time

from lib.sensors import OctopusDeploySensor


class NewReleaseSensor(OctopusDeploySensor):
    def __init__(self, sensor_service, config=None, poll_interval=None):
        super(NewReleaseSensor, self).__init__(
            sensor_service=sensor_service,
            config=config,
            poll_interval=poll_interval,
            trigger_ref='octopusdeploy.new_release',
            store_key='octopusdeploy.last_release_date_str')
        self._logger = self._sensor_service.get_logger(__name__)

    def poll(self):
        self._logger.debug('Requesting list of releases')
        releases = self._get_releases()

        # Make sure there are releases
        if releases is None:
            self._logger.info('No releases found')
            return
        if len(releases) == 0:
            self._logger.info('Empty list of releases')
            return

        last_release = releases[0]

        last_assembled_date = self._to_date(last_release['assembled'])

        # What is the last indexed release date? If none, index and exit
        index_date = self._get_last_date()
        self._logger.debug('Index date is %s' % index_date)
        if index_date is None:
            self._logger.info('Initializing index')
            self._set_last_date(last_assembled_date)
            index_date = self._get_last_date()

        # If there have been new releases, trigger them each
        if last_assembled_date > index_date:
            self._logger.info('Found releases to trigger')
            # Get releases since the last update time
            # They are in date order so once you get one behind the index
            # break out of the loop
            for release in releases:
                if self._to_date(release['assembled']) > index_date:
                    self._logger.info('Raising trigger for %s' % release['id'])
                    self._dispatch_trigger_for_payload(release)
                else:
                    break
            self._set_last_date(last_assembled_date)
        else:
            self._logger.debug('No new releases comparing %s'
                               % time.strftime("%Y-%m-%dT%H:%M:%S",
                                               last_assembled_date))
            self._logger.debug('and %s'
                               % time.strftime("%Y-%m-%dT%H:%M:%S",
                                               index_date))

    def _get_releases(self):
        result = self.make_get_request("releases")
        releases = self._to_triggers(result['Items'])
        return releases

    def _get_last_release(self):
        releases = self._get_releases()
        if releases is None:
            return None
        if len(releases) != 1:
            return None
        return releases[0]

    def _to_trigger(self, release):
        trigger = {
            'id': release['Id'],
            'assembled': release['Assembled'],
            'release_notes': release['ReleaseNotes'],
            'version': release['Version'],
            'author': release['LastModifiedBy'],
            'project_id': release['ProjectId']
        }
        return trigger


sensors/new_deployment_sensor.py
import time

from lib.sensors import OctopusDeploySensor


class NewDeploymentSensor(OctopusDeploySensor):
    def __init__(self, sensor_service, config=None, poll_interval=None):
        super(NewDeploymentSensor, self).__init__(
            sensor_service=sensor_service,
            config=config,
            poll_interval=poll_interval,
            trigger_ref='octopusdeploy.new_deployment',
            store_key='octopusdeploy.last_deploy_date_str')

        self._logger = self._sensor_service.get_logger(__name__)

    def poll(self):
        self._logger.debug('Requesting list of deployments')
        deployments = self._get_deployments()

        # Make sure there are releases
        if deployments is None:
            self._logger.info('No deployments found')
            return
        if len(deployments) == 0:
            self._logger.info('Empty list of deployments')
            return

        last_deployment = deployments[0]

        last_assembled_date = self._to_date(last_deployment['assembled'])

        # What is the last indexed release date? If none, index and exit
        index_date = self._get_last_date()
        self._logger.debug('Index date is %s' % index_date)
        if index_date is None:
            self._logger.info('Initializing index')
            self._set_last_date(last_assembled_date)
            index_date = self._get_last_date()

        # If there have been new deployments, trigger them each
        if last_assembled_date > index_date:
            self._logger.info('Found deployments to trigger')
            # Get deployments since the last update time
            # They are in date order so once you get one behind the index
            # break out of the loop
            for deployment in deployments:
                if self._to_date(deployment['assembled']) > index_date:
                    self._logger.info('Raising trigger for %s' % deployment['id'])
                    self._dispatch_trigger_for_payload(deployment)
                else:
                    break
            self._set_last_date(last_assembled_date)
        else:
            self._logger.debug('No new deployments comparing %s'
                               % time.strftime("%Y-%m-%dT%H:%M:%S",
                                               last_assembled_date))
            self._logger.debug('and %s'
                               % time.strftime("%Y-%m-%dT%H:%M:%S",
                                               index_date))

    def _get_deployments(self):
        result = self.make_get_request("deployments")
        releases = self._to_triggers(result['Items'])
        return releases

    def _get_last_deployment(self):
        deployments = self._get_deployments()
        if deployments is None:
            return None
        if len(deployments) != 1:
            return None
        return deployments[0]

    def _to_trigger(self, deployment):
        trigger = {
            'id': deployment['Id'],
            'name': deployment['Name'],
            'comments': deployment['Comments'],
            'assembled': deployment['Created'],
            'version': deployment['ReleaseId'],
            'author': deployment['LastModifiedBy'],
            'project_id': deployment['ProjectId']
        }
        return trigger


sensors/lib/sensors.py
try:
    import requests
except ImportError:
    message = ('Missing "requests", please install it using pip:\n'
               'pip install requests')
    raise ImportError(message)

import json
import time

from st2reactor.sensor.base import PollingSensor

__all__ = [
    'OctopusDeploySensor',
]


class OctopusDeploySensor(PollingSensor):
    def __init__(self,
                 sensor_service,
                 config,
                 poll_interval,
                 trigger_ref,
                 store_key):
        super(OctopusDeploySensor, self).__init__(sensor_service=sensor_service,
                                                  config=config,
                                                  poll_interval=poll_interval)
        self.client = self._init_client()
        self._last_date = None
        self._trigger_ref = trigger_ref
        self._store_key = store_key

    def _init_client(self):
        api_key = self._config['api_key']
        host = self._config['host']
        port = self._config['port']
        return OctopusDeployClient(api_key=api_key, host=host, port=port)

    def _to_date(self, date_string):
        date_string = date_string.split('.')[0]
        return time.strptime(date_string, '%Y-%m-%dT%H:%M:%S')

    def _build_uri(self):
        # big assumption but it'll cover 99% case,
        # as octopus runs https by default
        start = "http://" if self.client.port == 80 else "https://"
        return start + self.client.host + ":" + str(self.client.port) + "/api/"

    def _get_last_date(self):
        self._last_date = self._sensor_service.get_value(name=self._store_key)
        if self._last_date is None:
            return None
        return time.strptime(self._last_date, '%Y-%m-%dT%H:%M:%S')

    def _set_last_date(self, last_date):
        self._last_date = time.strftime('%Y-%m-%dT%H:%M:%S', last_date)
        self._sensor_service.set_value(name=self._store_key,
                                       value=self._last_date)

    def _to_triggers(self, items):
        triggers = []
        for item in items:
            triggers.append(self._to_trigger(item))
        return triggers

    def _to_trigger(self, item):
        # Implement to copy fields
        pass

    def _dispatch_trigger_for_payload(self, trigger_payload):
        trigger = self._trigger_ref
        self._sensor_service.dispatch(trigger=trigger, payload=trigger_payload)

    def make_post_request(self, action, payload):
        response = requests.post(self._build_uri() + action,
                                 data=json.dumps(payload), verify=False,
                                 headers=self.client.headers)
        return response.json()

    def make_get_request(self, action):
        response = requests.get(self._build_uri() + action,
                                verify=False,
                                headers=self.client.headers)
        return response.json()

    def setup(self):
        pass

    def cleanup(self):
        pass

    def add_trigger(self, trigger):
        pass

    def update_trigger(self, trigger):
        pass

    def remove_trigger(self, trigger):
        pass


class OctopusDeployClient(object):
    def __init__(self, api_key, host, port):
        self.api_key = api_key
        self.host = host
        self.port = port
        self.headers = {'X-Octopus-ApiKey': self.api_key,
                        'Content-type': 'application/json',
                        'Accept': 'text/plain'}
